[FIRRTL] Add PrefixModules pass

youngar opened the pull request:
This pass looks for modules annotated with the
`NestedPrefixModulesAnnotation` and prefixes the names of all modules
instantiated underneath it.  This pass will duplicate a module to rename
only the modules instantiated in the hiearchy rooted at the annotated
module.  If there are no remaining instances of the original module, it
will be removed.

The supported `firrtl.module` annotation is:
```
{
    class = "sifive.enterprise.firrtl.NestedPrefixModulesAnnotation",
    prefix = "MyPrefix_",
    inclusive = true
}
```

If `inclusive` is true, it will also apply the prefix to the annotated
module.

lattner commented on include/circt/Dialect/FIRRTL/InstanceGraph.h:
```diff
@@ -0,0 +1,148 @@
+namespace firrtl {
+
+class InstanceGraphNode {
+  using EdgeSetT = llvm::SmallVector<InstanceGraphNode *, 1>;
```
I'd go with just `EdgeSet` or `EdgeArray` or `EdgeList`, dropping the T.  The capitalization approach in MLIR eliminates the need for _t convention.

Why an inline size of 1?  These nodes are all heap allocated, and I doubt that 1 really captures most cases.

youngar replied:
Switched it to a std::vector using `EdgeVec`.

lattner commented on include/circt/Dialect/FIRRTL/InstanceGraph.h:
```diff
@@ -0,0 +1,148 @@
+
+  using const_iterator = EdgeSetT::const_iterator;
+  const_iterator begin() const { return instances.begin(); }
+  const_iterator end() const { return instances.end(); }
```
Just a suggestion, but I'd dump all const'ness support here, because it is meaningless because C++ doesn't have a mutation model that works for graphs (a const iterator still products pointers to mutable nodes).  The right thing is something like:

`using iterator = EdgeSetT::const_iterator;`

This prevents clients from mutating the edge list with the iterator (not something anyone should want) but still reflects that the dereferenced nodes are mutable.

lattner replied:
xref: https://mlir.llvm.org/docs/Rationale/UsageOfConst/

youngar replied:
I dropped all the weird constness stuff. I think I did it correctly with `using iterator = EdgeVec::const_iterator`, but I should go over it once more.

lattner commented on include/circt/Dialect/FIRRTL/InstanceGraph.h:
```diff
@@ -0,0 +1,148 @@
+
+/// This graph tracks modules and where they are instantiated. This is the
+/// inverse of the regular instance graph.
+class InstanceGraph {
```
I'm not sure how to parse:

> This is the inverse of the regular instance graph.

What is a regular instance graph?  Should this class be named "InverseInstanceGraph"?

youngar replied:
This code went through some iterations.. Looks like I need to clean it up.  I would say a regular graph is `module A instantiates module B`, while the inverse is `module B instantiated by module A`.

The graph right now is only tracking the information for inverse edges, but I tried to leave a big hole in the implementation so that the regular graph could be added.  If you look at the graph traits below, only the `llvm::Inverse` trait implemented.  This was done with the assumption that it would be faster to combine both graphs into 1 analysis, instead of building the graphs separately.

I can pretty easily add the forward edges + graph traits, it's just that I didn't need them for what I'm doing.  If there are no use-cases of the regular graph, then it would make a ton of sense to rename this to "InverseInstanceGraph".

I need to think a bit harder about what is needed here.

youngar replied:
Big update to how this is implemented: its now tracking all instance operations, and you can iterate the uses of a modules, as well as the instance ops in a module.  There are regular and inverse graph traits implemented.

lattner commented on include/circt/Dialect/FIRRTL/InstanceGraph.h:
```diff
@@ -0,0 +1,148 @@
+/// inverse of the regular instance graph.
+class InstanceGraph {
+
+  using NodeMapT = std::vector<std::unique_ptr<InstanceGraphNode>>;
```
Plz drop the T.  Also is this really a Map given it is a vector?

lattner commented on include/circt/Dialect/FIRRTL/InstanceGraph.h:
```diff
@@ -0,0 +1,148 @@
+  // pair<Operation*, InstanceGraphNode>.
+
+  /// The storage for graph nodes, with deterministic iteration.
+  std::vector<std::unique_ptr<InstanceGraphNode>> nodes;
```
Use your NodeMap typedef?

lattner commented on include/circt/Dialect/FIRRTL/InstanceGraph.h:
```diff
@@ -0,0 +1,148 @@
+
+// Provide constant graph traits for iterating the modules in inverse order.
+template <>
+struct GraphTraits<Inverse<const circt::firrtl::InstanceGraphNode *>> {
```
again I'd drop the const support.

lattner commented on include/circt/Dialect/FIRRTL/Passes.td:
```diff
@@ -174,4 +174,29 @@ def BlackBoxReader : Pass<"firrtl-blackbox-reader", "CircuitOp"> {
+    `NestedPrefixModulesAnnotation` and prefixes the names of all modules
+    instantiated underneath it.  This pass will duplicate a module to rename
+    only the modules instantiated in the hiearchy rooted at the annotated
+    module.  If there are no remaining instances of the original module, it
```
Random wording suggestion, instead of:

This pass will duplicate a module to rename  only the modules instantiated in the hiearchy rooted at the annotated   module.  

Maybe something like:

This pass will duplicated modules as necessary to given submodules unique names, ...

also typo hiearchy

lattner commented on lib/Dialect/FIRRTL/InstanceGraph.cpp:
```diff
@@ -0,0 +1,32 @@
+//
+//===----------------------------------------------------------------------===//
+#include "llvm/Support/Debug.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
```
Plz include InstanceGraph.h first.  You can put a blank line between the includes if clang-format isn't cooperating

lattner commented on lib/Dialect/FIRRTL/InstanceGraph.cpp:
```diff
@@ -0,0 +1,32 @@
+#include "llvm/Support/Debug.h"
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+
+using namespace mlir;
```
can you the the `mlir` using decl?

lattner commented on lib/Dialect/FIRRTL/Transforms/PrefixModules.cpp:
```diff
@@ -0,0 +1,177 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
```
nit: drop `./`

lattner commented on lib/Dialect/FIRRTL/Transforms/PrefixModules.cpp:
```diff
@@ -0,0 +1,177 @@
+#include "circt/Dialect/FIRRTL/InstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Support/LLVM.h"
+#include "llvm/ADT/DenseMap.h"
```
I suspect you can drop some of these #includes

lattner commented on lib/Dialect/FIRRTL/Transforms/PrefixModules.cpp:
```diff
@@ -0,0 +1,177 @@
+
+/// Get the PrefixInfo for a module. This reads the annotations on module.  If
+/// the module is not annotated, the prefix returned will be empty.
+PrefixInfo getPrefixInfo(FModuleOp module) {
```
Plz mark function static to shrink the scope of the anonymous namespace.  It gives better "local reasoning" when reading source code.

lattner commented on lib/Dialect/FIRRTL/Transforms/PrefixModules.cpp:
```diff
@@ -0,0 +1,177 @@
+
+/// This maps a FModuleOp to a list of all prefixes that need to be applied.
+/// When a module has multiple prefixes, it will be cloned for each one.
+using PrefixMap = DenseMap<StringRef, llvm::SmallSet<std::string, 1>>;
```
I'm not following all the logic here, but this is an extremely expensive representation for this.  Do you really need a SmallSet here instead of an array like thing?  If duplicates are possible, you could toss all the things into the array then sort/eliminate dupes before using it.

Also, do you need string here, or can you use a StringRef into the underlying attributes?

You might also consider StringMap<x> instead of DenseMap<StringRef, x>

youngar replied:
A Set is logically what I need here, but I could try to using a vector.  The way this works, it will try to insert a new prefix in to the list for each instance of a module. In the majority of cases, each module is getting renamed with a single prefix (i.e. set size of 1) so maybe I can do a linear scan on insertion.

lattner reviewed (approved):
Looking great

youngar commented:
We had an offline conversation about this pass today, and what needs to happen to make sure we can support the SiFive flows.  Right now, this pass as implemented only looks for the annotation on FModuleOps. It turns out the the NestedPrefix annotation can be attached to Modules and Instances, even non-local instances.  What I found is that for our immediate use case, we only need to support attaching this annotation on Modules, but it is likely that we will need to support instances.  I plan on adding support for local instance op annotations before merging this pass. AFAICT there is no need for supporting non-local annotations.

This triggered a larger discussion about potentially removing this pass and adding a feature to Chisel to accomplish what this pass is used for.  What this would look like is still undecided, but it could mean better support for Chisel packages, module-renaming, and de-duplication domains (i.e. only de-duplicate modules if they are in the same hierarchy).

lattner commented:
Sounds good.  I'd recommend pulling the InstanceGraph out to its own thing - this is critical infra on its own right.  I'd check out how to get IMConstProp and other IM passes to use it.

fabianschuiki commented:
What's the current state of this pass? Did we decide to drop this in favor of a Chisel-side solution in the end?

youngar commented:
This pass is held up because we were waiting for non-local annotation support and we were going to rewrite this from scratch based on that work. I think there was some idea that this could get folded into some other pass like structural dedup or annotaion-scattering or something, but it's been a while since we talked about this.

If we don't need non-local annotation support, and after addressing Chris' review to switch from SmallSet to SmallVector, this pass is good to go as is.

youngar commented:
We're going to go with this version without non-local annotation support.  I have updated it after a couple of months of neglect:
1. Rebased, the InstanceGraph part of this PR was merged a long time ago.
2. It now prefixes memory names.
3. Enabled by default in firtool.
4. Support for `markAllAnalysisPreserved()`.
5. Switched from a `SmallSet` to a `SmallVector` for recording required prefixes per module.
6. Tweaked the order of cloned modules in the output so they are generated after the reference module.
7. Small code tweaks.

seldridge commented on tools/firtool/firtool.cpp:
```diff
@@ -270,6 +275,9 @@ processBuffer(MLIRContext &context, TimingScope &ts, llvm::SourceMgr &sourceMgr,
   if (inferResets)
     pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInferResetsPass());
 
+  if (prefixModules)
+    pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInferResetsPass());
```
Should be prefix modules. 😅

youngar replied:
🤦

seldridge commented on test/Dialect/FIRRTL/prefix-modules.mlir:
```diff
@@ -0,0 +1,108 @@
+}
+
+// CHECK: firrtl.module @T_Zebra
+// CHECK-NOT: firrtl.module @T_Zebra
```
This comment was super old when I started a review. Resolving this in favor of the up-to-date one.

seldridge commented:
I'm totally good with punting on non-local annotations here. Is there some check that we can apply to warn/error out if we see those at this pass?

youngar commented on include/circt/Dialect/FIRRTL/Passes.td:
````diff
@@ -218,6 +218,30 @@ def BlackBoxReader : Pass<"firrtl-blackbox-reader", "CircuitOp"> {
+    definitions, as well as specific instances.
+
+    The supported annotation is:
+    ```
+      {
+        class = "sifive.enterprise.firrtl.NestedPrefixModulesAnnotation",
+        prefix = "MyPrefix_",
+        inclusive = true
+      }
+    ```
````
It's already there :)

youngar commented on lib/Dialect/FIRRTL/Transforms/PrefixModules.cpp:
```diff
@@ -0,0 +1,211 @@
+static PrefixInfo getPrefixInfo(Operation *module) {
+  AnnotationSet annotations(module);
+
+  // Get the annotation from the module.
+  auto dict = annotations.getAnnotation(
+      "sifive.enterprise.firrtl.NestedPrefixModulesAnnotation");
+  if (!dict)
+    return {"", false};
+  Annotation anno(dict);
```
Yeah this is surprising.  I did a quick check if there was anything there and it looks like they all return DictionaryAttrs.

youngar replied:
Yeah this is surprising. I did a quick check if there was anything there and it looks like they all return DictionaryAttrs.

youngar commented on include/circt/Dialect/FIRRTL/Passes.td:
````diff
@@ -218,6 +218,30 @@ def BlackBoxReader : Pass<"firrtl-blackbox-reader", "CircuitOp"> {
+    ```
+
+    If `inclusive` is false, it will not attach the prefix to target module,
+    only to modules instantiated underneath it.
````
I updated it to be identical to the text from the Annotation document, it covers both cases.  It's still worded a little weirdly though

youngar commented:
> Is there some check that we can apply to warn/error out if we see those at this pass?
This comment made me realize that I'm not actually removing the annotations, which I have updated to fix.

Is it good enough that these will be caught by `warn-unprocessed-annotations`?

youngar commented:
> Is it good enough that these will be caught by warn-unprocessed-annotations?

Offline conversation: this is good enough for now, we can't detect all the things people can do wrong with annotations.  Eventually @darthscsi will be adding the ability to verify specific annotations when we attach them to the IR and we can give better error messages there.
